Human v human

rianp opened the pull request:
## Changes made:

1. Add update board functionality
2. add change turn functionality

## Acceptance criteria:

1. The game should be played between two human players. Assign player 1 as 'X’ and player 2 as 'O.’
2. For each turn, print the board and a message stating whose turn it is.
3. Players can mark available cells with their marker.
4. If a player attempts to mark an invalid cell, then show a corresponding message and allow the player to retry their turn.

[ticket](https://trello.com/c/FF5OqmWv)

Commit:
add update board functionality
  - add change turn functionality

rianp committed:
move board change state and change turn functionality
  - rename test files to jsx

rianp committed:
add tests for updateBoard and changeTurn functionality

rianp committed:
move turn calculation to their own functions
  - refactor tests

ranizilpelwar commented on src/board/Board.jsx:
```diff
@@ -1,20 +1,32 @@
+    setBoardData(updatedBoard);
+
+    changeTurn();
+  };
```
This works here, you are using internal state to remember the board and calling the prop when it is done.

ranizilpelwar commented on src/board/Board.test.jsx:
```diff
@@ -0,0 +1,42 @@
+    fireEvent.click(cellOne);
+    expect(screen.queryByText("1")).not.toBeInTheDocument();
+    expect(cellOne).toHaveTextContent("z");
+  });
```
would it be worth making sure the other cells didn't get updated?

ranizilpelwar commented on src/board/Cell.test.jsx:
```diff
@@ -0,0 +1,25 @@
+
+  const cell = getByText("abc");
+  userEvent.click(cell);
+
```
I believe the best practice is using screen.getByText instead of pulling the functions from the render.

ranizilpelwar commented on src/changeTurn.js:
```diff
@@ -0,0 +1,17 @@
+
+  const currentPlayerMark = currentPlayerId === 0 ? "x" : "o";
+  return currentPlayerMark;
+}
```
this is cool, nice work

ranizilpelwar reviewed (approved):
I like the refactor and your new approach. I like that you kept the useState need where it was needed. Would be cool if you get to useEffect on a component sometime later but use it as a "when needed" approach.

rianp committed:
add turn display
  - implement edge case handling for updateBoard and changeTurn

rianp committed:
add test for unclicked cells remain unchanged

rianp committed:
add move already taken alert

rianp committed:
add close alert functionality

clairenguyen commented on src/App.test.jsx:
```diff
@@ -0,0 +1,34 @@
+test("displays the Tic-Tac-Toe header", () => {
+  render(<App />);
+  const headerElement = screen.getByText(/Tic-Tac-Toe/i);
+  expect(headerElement).toBeInTheDocument();
```
can you change any tests where you're testing what a user sees to be  `.toBeVisible()`?  

generally you'd use  `.toBeInTheDocument()`  when you're checking something is not there (e.g. `queryByText("something").not.toBeInTheDocument()`, or the thing you want to assert on might actually be in the HTML/document but might be hidden.

rianp replied:
awesome!

clairenguyen commented on src/App.test.jsx:
```diff
@@ -0,0 +1,34 @@
+  const alertMessage = screen.getByText(
+    "Move already made. Please choose an empty cell."
+  );
+  expect(alertMessage).toBeInTheDocument();
```
.toBeVisible()

clairenguyen commented on src/alert/Alert.jsx:
```diff
@@ -0,0 +1,15 @@
+      </span>
+    </div>
+  );
+};
```
I noticed you didn't include prop types here but they're in the other components. curious if your IDE indicated anything to you or not.

rianp replied:
My IDE didn't indicate anything. Also when I moved updateGame functionality into the Game component, my console threw an error where it wanted me to get rid of prop types for the Board component, so I refactored those out, and everything is fine now.

clairenguyen commented on src/alert/Alert.test.jsx:
```diff
@@ -0,0 +1,21 @@
+describe("Alert Component", () => {
+  it("should render the message correctly", () => {
+    const message = "Test message";
+    render(<Alert message={message} />);
```
If your prop types were included in your Alert.jsx, I'd imagine it would indicate that your onClose prop is missing here. Unless you intentionally want onClose to be an optional prop.

rianp replied:
I refactored out onClose. But I'll look into this just to double check I'm not missing anything :)

clairenguyen commented on src/board/updateBoard.test.js:
```diff
@@ -0,0 +1,49 @@
+import { updateBoard } from "./updateBoard"; // Replace "./your-functions-file" with the actual path to your functions file.
```
did you mean to leave this comment here?

rianp replied:
no. thank you for the catch!

clairenguyen commented on src/changeTurn.js:
```diff
@@ -0,0 +1,17 @@
+export function calculateCurrentPlayerId(currentPlayerId) {
+  if (currentPlayerId !== 0 && currentPlayerId !== 1) {
+    throw new Error("Invalid currentPlayerId. It should be either 0 or 1.");
```
nice error checking

clairenguyen commented on src/changeTurn.test.js:
```diff
@@ -0,0 +1,60 @@
+
+  it("should throw an error message when currentPlayerId is not 0 or 1", () => {
+    const currentPlayerId = -3;
+    expect(() => calculateCurrentPlayerMark(currentPlayerId)).toThrowError(
```
i'm curious as to why the expect is in a callback function. Is that necessary? just wondering because i've never written it as a call back before. I'd usually write it like..

```
expect(calculateCurrentPlayerMark(currentPlayerId)).toThrowError() 
```

clairenguyen commented on src/game/updateGame.js:
```diff
@@ -0,0 +1,21 @@
+export const updateGame = (
```
don't need to address this now but this feel like a bit of a code smell with this many arguments/parameters.  wondering if this instead of passing in boardData, cellValue, and currentPlayer, can you just pass in the already calculated "updateBoard" value?

clairenguyen replied:
also, i'm wondering if there's a reason why you chose to pass in `updateBoard` as a parameter instead of importing it in to the file?

rianp replied:
hmm. I could do this. But wouldn't updateBoard get updated regardless of the validity of the move in this case? Wouldn't this create a lot of unnecessary actions? Maybe I am missing something.

I definitely don't need to pass in updateBoard as a param. thank you for the catch on this.

clairenguyen commented on src/board/Board.jsx:
```diff
@@ -1,20 +1,39 @@
+  const [boardData, setBoardData] = useState(defaultBoardData);
+
+  const handleClick = (cellValue) => {
+    const updatedBoard = updateGame(
```
i mentioned this in the file itself, but this could probably get refactored at a later point. feels like a potential code smell with so many parameters, including the `updateBoard` method

rianp replied:
I refactored updatedBoard variable out :)

clairenguyen commented on src/game/updateGame.js:
```diff
@@ -0,0 +1,21 @@
+    showAlert();
+  }
+
+  return updatedBoard;
```
A couple refactor suggestions for later (I'm writing them down as they come to me to let you see my thought process of how I might refactor this) :

First, I would try to avoid using `let` if possible. But if you're going to use it, maybe set `updatedBoard = boardData` inside the `else` to better communicate that you want the original board? i actually missed it when you initialized it on line 10. e.g. 

```
let updatedBoard;

  if (cellValue !== "X" && cellValue !== "O") {
    updatedBoard = updateBoard(boardData, cellValue, currentPlayer);
    closeAlert();
    changeTurn();
  } else {
    updatedBoard = boardData;
    showAlert();
  }

  return updatedBoard;
```

clairenguyen replied:
other potential rewrite: 

```
  if (cellValue === "X" && cellValue === "O") {
    showAlert()
    return boardData
  } else {
    closeAlert();
    changeTurn();
    return updateBoard(boardData, cellValue, currentPlayer);
  }
```
tbh, not quite a fan of this, but it gets rid of the `let` and communicates that the return value for each condition is different.

clairenguyen replied:
Last suggestion, what if you move the responsibility to use the original board vs updating the board directly into updateBoard? e.g. 

updateGame.js:
```
 if (cellValue === "X" && cellValue === "O") {
    showAlert()
  } else {
    closeAlert();
    changeTurn();
  }
  
  return updateBoard(boardData, cellValue, currentPlayer);
```

updateBoard.js:
```
export function updateBoard(board, value, turn) {

if (value !== "X" && value !== "O") {
  return board
}

return board.map((row) =>
    row.map((cell) => (cell === value ? turn : cell))
  );
}
```

if you do this, then I think you can just call`updateBoard()` directly in Board.jsx > handleClick() instead of passing it down into `updateGame()`. It just feels like `updateGame` is doing a lot.

clairenguyen replied:
just some ideas. curious if @ranizilpelwar  or @fifikim have thoughts on this.

rianp replied:
I know what you mean. It is a lot of parameters. After going over many different refactor options on this, I am still trying to figure out how to minimize the params being passed into updateGame. Most of this is because I'm keeping logic out of the Game component. If I weren't doing this, there would be no need to call an updateGame function or pass so many parameters into it. I am wondering if you have any other suggestions after this most recent refactor.

ranizilpelwar commented on src/alert/Alert.jsx:
```diff
@@ -0,0 +1,15 @@
+    </div>
+  );
+};
+
```
neat, reusable component. things like this can be in a "common/components" folder if they are used in other places, generally in a larger app. just posting this here so you know :)

ranizilpelwar commented on src/App.jsx:
```diff
@@ -14,8 +34,23 @@ export default function App() {
+        <Board
+          showAlert={showAlert}
+          closeAlert={closeAlert}
+          changeTurn={changeTurn}
```
without seeing the UI in action and digging deep into the details, can I ask, why do you need to pass in the show/close alert into the updateGame function if the Alert is shown above here?

rianp replied:
During my check-in with Sophia yesterday, Sophia suggested it might be cool to have the alert close automatically when the user hits an available square in addition to having a close alert button. This made it necessary to pass close alert through. showAlert is passed through because it is triggered when a user hits an unavailable cell.

clairenguyen commented on src/board/Board.test.jsx:
```diff
@@ -0,0 +1,81 @@
+});
+
+describe("when user clicks a cell", () => {
+  test("board is updated with correct marker", () => {
```
Small detail but I'd maybe adjust some of the spacing here to improve readability for the next developer. Every team/test is different but I generally like to keep my "action" (e.g. the .click() in a separate block) from the expect and/or the setup. It depends how many lines of code is already in the test / how readable it already is. for example, for this test, i might write this as: 

```   
test("board is updated with correct marker", () => {
  render(
      <Board
        showAlert={jest.fn()}
        closeAlert={jest.fn()}
        changeTurn={jest.fn()}
        currentPlayer="z"
      />
    );
   
   const cellOne = screen.getByText("1");
   fireEvent.click(cellOne);

   expect(screen.queryByText("1")).not.toBeInTheDocument();
   expect(cellOne).toHaveTextContent("z");
})
```
Not a huge deal; it just felt off that `const cellOne = screen.getByText("1");` was isolated and it drew my attention to that line

Also, you don't need variables for every mocked function (e.g. in your props) unless you're planning on asserting it gets called, or to add some clarity with the variable names.

ranizilpelwar commented on src/game/updateGame.js:
```diff
@@ -0,0 +1,21 @@
+    changeTurn();
+  } else {
+    showAlert();
+  }
```
To continue my previous Q, perhaps you are passing in show/close alert as a way to "do something when there is an invalid state". If so, I wonder if you can decouple the concept of showing the alert from the concept of an invalid state. An initial idea is to rename the callbacks so they are not based on the UI, so if the UI changes later to show something else (instead of an alert), your interface is the same. ex: onValidationFailure() is a name of a function that I can think of, preliminarily. If this doesn't make sense for what you were trying to achieve, no worries, just ignore this comment.

ranizilpelwar replied:
One more thing to consider, I see that updateGame happens inside `<Board/>`. Is this the optimal place for it? Would board control how the game is played in addition to how it is rendered? 
This isn't a deal breaker for me for PR approval, but of all the comments I added, I think this is the one that I would like you to see if you can find an alternative solution to. It seems like Board is doing too much. Can you think of how data flows in your app and what could be missing here?

ranizilpelwar replied:
Also in functional programming, the pattern is to not mutate a value; ie "let" is avoided. Can you see how to refactor it to use const instead? Not a deal breaker for the PR but something you want to get practice in.

rianp replied:
Awesome. This is good advice. I'll work on moving the updateGame functionality out of Board and into a Game component that calls Board. I think this would make a lot more sense and maybe minimize the amount of props being passed into Board.

I definitely see what you mean about showAlert/closeAlert. I'll refactor this into a validator of some kind. Thank you!

ranizilpelwar commented on src/game/updateGame.test.js:
```diff
@@ -0,0 +1,82 @@
+    ["O", "8", "X"],
+  ];
+
+  it('should update the board and change turn if the cellValue is not "X" or "O"', () => {
```
When I read this, I would have liked to know what the significance of not X or 0 was. I didn't initially know. 
I think it is for an unmarked cell. If so, the test title would have been better served, as split like the following: describe("when the cell is unmarked") and it("should update....")
Also I think you could have two tests in that describe, one for "update the board" and one for "change turn". The pattern is usually to have a single expect per test but sometimes more if they are really related. 

Another reason to decouple the test name from the values you enter in the cells is when the board changes to emojis for example, your test title can help you know that the behavior to test is that if the cell is not already marked, do this. It doesn't matter what the mark is. When it changes to emojis, you can change the test arrange section to use that new value instead.

rianp replied:
oh this is good advice thank you!

clairenguyen commented on src/board/Board.test.jsx:
```diff
@@ -0,0 +1,81 @@
+    expect(cellOne).toHaveTextContent("z");
+  });
+
+  test("changes turn", () => {
```
to follow up on my previous comment,  i'd only have one variable for the mocked function that you want to test. it keeps the focus on `changeTurn` and indicates that it doesn't really matter what the values of the other props are.   
```
  const mockChangeTurn = jest.fn();

    render(
      <Board
        showAlert={jest.fn()}
        closeAlert={jest.fn()}
        changeTurn={mockChangeTurn}
        currentPlayer="z"
      />
    );

    const cellOne = screen.getByText("1");
    fireEvent.click(cellOne);
    
    expect(mockChangeTurn).toHaveBeenCalled();
  });
  ```
  
Part of the reason why i'm mentioning this is because for my current client code base, there is an excessive amount of variables set to jest.fn() but they don't do anything; they end up being more of a distraction and it's hard to know which part actually matters

clairenguyen reviewed (changes requested):
Overall good job! I wrote a lot of comments/suggestions on future improvements and some nice-to-knows about react/testing.  Don't need to address them all now (specifically the large comment on updateGame).  The other ones should be smaller/quick fixes. You clearly have learned a lot about react and javascript during this project so far!

clairenguyen commented on src/App.jsx:
```diff
@@ -14,8 +34,23 @@ export default function App() {
+        <div>
+          {isAlertVisible && (
+            <Alert
+              message="Move already made. Please choose an empty cell."
```
This `message` prop seems to be a pre-optimization. This is the only place that `Alert` is being used at the moment.  I would move the message text into Alert until you actually need it to be a prop.

rianp replied:
oh true! good point.

rianp committed:
move update game functionality
  - refactor alert

rianp committed:
make board and status bar prettier

rianp committed:
pr comment refactors

rianp committed:
refactor for params smell
  - move updateBoard into Board
  - implement validate changeTurn
  - delete updateGame

rianp committed:
delete Game component

rianp committed:
refactor and rename variables, props, and tests
